### ISSUE TITLE
Broadcom 4354 chip doesn't work

### ISSUE THREAD

**Issue by denysvitali** [denysvitali/linux-on-pixel-c#3]
Currently the Bluetooth / Wi-Fi chip doesn't seem to power on, my assumption is that there is a GPIO control somewhere

**Comment by bushvin** [comment]
Have you tried rfkill?

**Comment by denysvitali** [comment]
The wifi chip doesn't show up in the [dmesg](https://github.com/denysvitali/linux-on-pixel-c/blob/master/dmesg/4.14-rc4_20171006_094829.md), looks like it doesn't get activated (enable pin set to LOW?) nor in `ip link` / `ifconfig`.  
  
I may have found the source of the problem though: [we're missing some consumers](https://github.com/denysvitali/linux-smaug/blob/4.14-rc6/arch/arm64/boot/dts/nvidia/tegra210-smaug-sysedp.dtsi#L403-L407), that are in [Google's tegra210-smaug-sysedp.dtsi](https://android.googlesource.com/kernel/tegra/+/android-tegra-dragon-3.18-oreo-r6/arch/arm64/boot/dts/tegra/tegra210-smaug-sysedp.dtsi)

**Comment by denysvitali** [comment]
The first problem was that the module Kconfig file had a different name, therefore it didn't get compiled. Now we have to port the brcmdhd driver to 4.11.

These are the errors, as of today:
```
drivers/net/wireless/bcmdhd/dhd_linux.c: In function 'dhd_watchdog_thread':
drivers/net/wireless/bcmdhd/dhd_linux.c:3211:22: error: storage size of 'param' isn't known
   struct sched_param param;
                      ^
drivers/net/wireless/bcmdhd/dhd_linux.c:3211:22: warning: unused variable 'param' [-Wunused-variable]
drivers/net/wireless/bcmdhd/dhd_linux.c: In function 'dhd_dpc_thread':
drivers/net/wireless/bcmdhd/dhd_linux.c:3328:22: error: storage size of 'param' isn't known
   struct sched_param param;
                      ^
drivers/net/wireless/bcmdhd/dhd_linux.c:3328:22: warning: unused variable 'param' [-Wunused-variable]
drivers/net/wireless/bcmdhd/dhd_linux.c: In function 'dhd_rxf_thread':
drivers/net/wireless/bcmdhd/dhd_linux.c:3389:22: error: storage size of 'param' isn't known
   struct sched_param param;
                      ^
drivers/net/wireless/bcmdhd/dhd_linux.c:3389:22: warning: unused variable 'param' [-Wunused-variable]
drivers/net/wireless/bcmdhd/dhd_linux.c: In function 'dhd_allocate_if':
drivers/net/wireless/bcmdhd/dhd_linux.c:4436:11: error: 'struct net_device' has no member named 'destructor'
   ifp->net->destructor = free_netdev;
           ^
drivers/net/wireless/bcmdhd/dhd_linux.c:4438:11: error: 'struct net_device' has no member named 'destructor'
   ifp->net->destructor = dhd_netdev_free;
           ^
  CC      drivers/net/wireless/bcmdhd/hndpmu.o
drivers/net/wireless/bcmdhd/dhd_linux.c: In function 'dhd_os_get_image_block':
drivers/net/wireless/bcmdhd/dhd_linux.c:7302:26: warning: passing argument 2 of 'kernel_read' makes pointer from integer without a cast
  rdlen = kernel_read(fp, fp->f_pos, buf, len);
                          ^
In file included from ./include/linux/huge_mm.h:6:0,
                 from ./include/linux/mm.h:451,
                 from drivers/net/wireless/bcmdhd/include/linuxver.h:72,
                 from drivers/net/wireless/bcmdhd/dhd_linux.c:29:
./include/linux/fs.h:2822:16: note: expected 'void *' but argument is of type 'loff_t'
 extern ssize_t kernel_read(struct file *, void *, size_t, loff_t *);
                ^
drivers/net/wireless/bcmdhd/dhd_linux.c:7302:37: warning: passing argument 3 of 'kernel_read' makes integer from pointer without a cast
  rdlen = kernel_read(fp, fp->f_pos, buf, len);
                                     ^
In file included from ./include/linux/huge_mm.h:6:0,
                 from ./include/linux/mm.h:451,
                 from drivers/net/wireless/bcmdhd/include/linuxver.h:72,
                 from drivers/net/wireless/bcmdhd/dhd_linux.c:29:
./include/linux/fs.h:2822:16: note: expected 'size_t' but argument is of type 'char *'
 extern ssize_t kernel_read(struct file *, void *, size_t, loff_t *);
                ^
drivers/net/wireless/bcmdhd/dhd_linux.c:7302:42: warning: passing argument 4 of 'kernel_read' makes pointer from integer without a cast
  rdlen = kernel_read(fp, fp->f_pos, buf, len);
                                          ^
In file included from ./include/linux/huge_mm.h:6:0,
                 from ./include/linux/mm.h:451,
                 from drivers/net/wireless/bcmdhd/include/linuxver.h:72,
                 from drivers/net/wireless/bcmdhd/dhd_linux.c:29:
./include/linux/fs.h:2822:16: note: expected 'loff_t *' but argument is of type 'int'
 extern ssize_t kernel_read(struct file *, void *, size_t, loff_t *);
                ^
drivers/net/wireless/bcmdhd/dhd_linux.c: In function 'dhd_set_version_info':
drivers/net/wireless/bcmdhd/dhd_linux.c:9148:3: error: implicit declaration of function 'dhd_bus_chip_id' [-Werror=implicit-function-declaration]
   "\n  Chip: %x Rev %x Pkg %x", dhd_bus_chip_id(dhdp),
   ^
drivers/net/wireless/bcmdhd/dhd_linux.c:9149:3: error: implicit declaration of function 'dhd_bus_chiprev_id' [-Werror=implicit-function-declaration]
   dhd_bus_chiprev_id(dhdp), dhd_bus_chippkg_id(dhdp));
   ^
drivers/net/wireless/bcmdhd/dhd_linux.c:9149:3: error: implicit declaration of function 'dhd_bus_chippkg_id' [-Werror=implicit-function-declaration]
cc1: some warnings being treated as errors
scripts/Makefile.build:313: recipe for target 'drivers/net/wireless/bcmdhd/dhd_linux.o' failed
make[4]: *** [drivers/net/wireless/bcmdhd/dhd_linux.o] Error 1
make[4]: *** Waiting for unfinished jobs....
drivers/net/wireless/bcmdhd/bcmwifi_channels.c: In function 'wf_chspec_aton':
drivers/net/wireless/bcmdhd/bcmwifi_channels.c:353:2: error: implicit declaration of function 'tolower' [-Werror=implicit-function-declaration]
  c = tolower((int)a[0]);
  ^
cc1: some warnings being treated as errors
scripts/Makefile.build:313: recipe for target 'drivers/net/wireless/bcmdhd/bcmwifi_channels.o' failed
make[4]: *** [drivers/net/wireless/bcmdhd/bcmwifi_channels.o] Error 1
scripts/Makefile.build:572: recipe for target 'drivers/net/wireless/bcmdhd' failed
make[3]: *** [drivers/net/wireless/bcmdhd] Error 2
scripts/Makefile.build:572: recipe for target 'drivers/net/wireless' failed
make[2]: *** [drivers/net/wireless] Error 2
make[2]: *** Waiting for unfinished jobs....
scripts/Makefile.build:572: recipe for target 'drivers/net' failed
make[1]: *** [drivers/net] Error 2
make[1]: *** Waiting for unfinished jobs....
Makefile:1023: recipe for target 'drivers' failed
make: *** [drivers] Error 2
```

**Comment by opendata26** [comment]
@denysvitali bcmdhd shouldn't be needed, the brcmfmac driver should work fine

**Comment by denysvitali** [comment]
Actually, the problem may be [this](https://github.com/denysvitali/linux-smaug/commit/6b35925e35fdccd181833e52c99ca265f9d64dd1#commitcomment-25445634). Yes, I'm stupid.

**Comment by denysvitali** [comment]
It still doesn't work. Apparently the WiFi chip needs to be enabled via a GPIO, but I still haven't figured out how to do it.

cc/ @samt43 @vartom

**Comment by vartom** [comment]
Enabled WiFi, In the kernel logs it now asks for firmware. I did not manage to find the working firmware yet. ps Firmware from rootfs, he does not see.
https://github.com/vartom/linux-smaug/commit/becac257eb89e443c5e9fa9a1eadf926987566cf

> [    0.955001] mmc1: new high speed SDIO card at address 0001
> [    0.966982] brcmfmac: brcmf_fw_map_chip_to_name: using brcm/brcmfmac4354-sdio.bin for chip 0x004354(17236) rev 0x000001
> [    0.967194] brcmfmac mmc1:0001:1: Direct firmware load for brcm/brcmfmac4354-sdio.bin failed with error -2

add firmware from linux firmware lib
[    0.945638] mmc1: new high speed SDIO card at address 0001
[    0.953578] brcmfmac: brcmf_fw_map_chip_to_name: using brcm/brcmfmac4354-sdio.bin for chip 0x004354(17236) rev 0x000001
[    1.197976] brcmfmac mmc1:0001:1: Direct firmware load for brcm/brcmfmac4354-sdio.clm_blob failed with error -2
[    1.197981] brcmfmac: brcmf_c_process_clm_blob: no clm_blob available(err=-2), device may have limited channels available
[    1.198497] brcmfmac: brcmf_c_preinit_dcmds: Firmware version = wl0: Jun  5 2017 08:54:22 version 7.35.101.6 (r702795) FWID 01-1fb17091

**Comment by denysvitali** [comment]
Is the chip enabled? The problem w/ the Broadcom drivers for the Pixel C is that they manually enable some GPIOs in order to activate / reset the chip.  
The mainline broadcom driver doesn't seem to work and porting Google's 3.18 brcmdhd driver to mainline seems an intensive task.
https://android.googlesource.com/kernel/tegra/+/android-8.1.0_r0.23/drivers/net/wireless/bcmdhd/
CC/ @samt43

**Comment by vartom** [comment]
I do not think that adding a driver from 3.18 is a good idea. If the built-in driver does not work, it's best to try the driver from 4.4 from NVIDIA

**Comment by denysvitali** [comment]
Indeed. We need to find a solution to integrate the chip power on feature in the broadcom mainline drivers.
By the way, you can download the needed firmware files by using my ./getvendor.sh script from the 4.15 branch

**Comment by vartom** [comment]
WiFi works. In your Linux build through the interface is not defined. In the assembly @Samt43 WiFi connects and works. 
Nothing new was added to the DT. Only added firmware.

**Comment by denysvitali** [comment]
Can you provide some instructions / make a PR to the tegra-nouveau-rootfs repo to enable it?

**Comment by Samt43** [comment]
@Vartom I never tried to make it work until now... But you mean it's
working with my courrent rootfs on GitHub ? Did you use 4.8 kernel ?
That's great !

Le jeu. 15 févr. 2018 17:00, Denys Vitali <notifications@github.com> a
écrit :

**Comment by vartom** [comment]
I mean the system from the first message of the[ topic](https://forum.xda-developers.com/pixel-c/orig-development/arch-running-natively-pixel-c-t3586368)
[Kernel 4.15.2](https://forum.xda-developers.com/showpost.php?p=75577446&postcount=41) from my repo.

**Comment by Samt43** [comment]
That's great !
I suppose the only thing to do then is to fix the current built rootfs by installing networkmanager (if i remember that was the one i installed some time ago). When creating rootfs, please use my generaterootfs script to fix suid correctly.
On kernel side, everything this to be there too ! :)

**Comment by denysvitali** [comment]
I'll take a look at this and fix the rootfs script accordingly :)
@vartom where are the firnware files for the BCM4354 module, and where did you placed them? Did you include them on the kernel or you've added them to the rootfs?

**Comment by vartom** [comment]
Completed the necessary patches in the repo.

**Comment by denysvitali** [comment]
Can you make a PR?

**Comment by vartom** [comment]
I began to notice problems in Android. This began after the inclusion of the WiFI in the kernel and the entry into Arch Linux(Samt43). To fix the problem, I cleaned Dalvik Cash. Something bad is happening to the file system.

**Comment by Samt43** [comment]
Hi @vartom ,
I compiled your 4.15 kernel and run my pixel C on it, compilation with tegra210 config was fine.
I couldn't have wlan node created unfortunatly due to this : 
[    1.194287] brcmfmac mmc1:0001:1: Direct firmware load for brcm/brcmfmac4354-sdio.clm_blob failed with error -2
It seems that firmware couldn't be loaded, so i checked CONFIG_FIRMWARE_IN_KERNEL=y and CONFIG_EXTRA_FIRMWARE and it seems that everything in correct to have firmware built in kernel.
As it works for you, im wondering if you put your firmware in initrd too.
Any idea ?!
Thanks !

**Comment by Samt43** [comment]
Sorry for this, it seems that brcmfmac4354-sdio.clm_blob is optional according to kernel code.
I will recheck tonight why i couldn't find any wlan reference in dmesg. Maybe everything was working !

**Comment by vartom** [comment]
All right brcmfmac4354-sdio.clm_blob is optional.

**Comment by denysvitali** [comment]
It works.  
I'll fix my initramfs accordingly and the rootfs.  

![photo_2018-02-19_13-56-27](https://user-images.githubusercontent.com/4939519/36379101-a3962fba-157d-11e8-8ae5-2435d8b2fbf1.jpg)

**Comment by vartom** [comment]
@denysvitali  Which version of rootfs you use?

**Comment by denysvitali** [comment]
I'm using a custom one I built - which is very similar to the latest build available, I'll push the changes in a few minutes

**Comment by denysvitali** [comment]
Unfortunately Bluetooth doesn't work yet, even though it appears in the rfkill list and in `/sys/class/bluetooth/hci0`

```
[root@alarm hci0]# ls -la
total 0
drwxr-xr-x 4 root root    0 Feb 19 14:44 .
drwxr-xr-x 3 root root    0 Feb 19 14:44 ..
lrwxrwxrwx 1 root root    0 Feb 19 14:59 device -> ../../../mmc1:0001:3
drwxr-xr-x 2 root root    0 Feb 19 14:59 power
drwxr-xr-x 3 root root    0 Feb 19 14:44 rfkill0
lrwxrwxrwx 1 root root    0 Feb 19 14:44 subsystem -> ../../../../../../../../../class/bluetooth
-rw-r--r-- 1 root root 4096 Feb 19 14:44 uevent
```

```
[root@alarm ~]# rfkill
ID TYPE      DEVICE      SOFT      HARD
 0 bluetooth hci0   unblocked unblocked
 1 wlan      phy0   unblocked unblocked
```

```
Feb 19 14:44:34 alarm systemd-udevd[224]: created db file '/run/udev/data/+bluetooth:hci0' for '/devices/platform/700b0200.sdhci/mmc_host/mmc1/mmc1:0001/mmc1:0001:3/bluetooth/hci0'
Feb 19 14:44:34 alarm systemd-udevd[224]: created db file '/run/udev/data/+rfkill:rfkill0' for '/devices/platform/700b0200.sdhci/mmc_host/mmc1/mmc1:0001/mmc1:0001:3/bluetooth/hci0/rfkill0'
Feb 19 14:44:34 alarm systemd[1]: sys-subsystem-bluetooth-devices-hci0.device: Changed dead -> plugged
Feb 19 14:44:34 alarm systemd[1]: sys-devices-platform-700b0200.sdhci-mmc_host-mmc1-mmc1:0001-mmc1:0001:3-bluetooth-hci0.device: Changed dead -> plugged
Feb 19 14:44:34 alarm systemd[1]: bluetooth.target: Trying to enqueue job bluetooth.target/start/fail
Feb 19 14:44:34 alarm systemd[1]: bluetooth.target: Installed new job bluetooth.target/start as 106
Feb 19 14:44:34 alarm systemd[1]: bluetooth.service: Installed new job bluetooth.service/start as 107
Feb 19 14:44:34 alarm systemd[1]: bluetooth.target: Enqueued job bluetooth.target/start as 106
Feb 19 14:44:35 alarm systemd-rfkill[250]: A new rfkill device has been added with index 0 and type bluetooth.
Feb 19 14:44:35 alarm systemd-rfkill[250]: Loaded state '0' from /var/lib/systemd/rfkill/platform-700b0200.sdhci:bluetooth.
Feb 19 14:44:37 alarm systemd[1]: bluetooth.service: ConditionPathIsDirectory=/sys/class/bluetooth succeeded.
Feb 19 14:44:37 alarm systemd[1]: bluetooth.service: Passing 0 fds to service
Feb 19 14:44:37 alarm systemd[1]: bluetooth.service: About to execute: /usr/lib/bluetooth/bluetoothd
Feb 19 14:44:37 alarm systemd[1]: bluetooth.service: Forked /usr/lib/bluetooth/bluetoothd as 260
Feb 19 14:44:37 alarm systemd[1]: bluetooth.service: Changed dead -> start
Feb 19 14:44:37 alarm systemd[260]: bluetooth.service: Executing: /usr/lib/bluetooth/bluetoothd
Feb 19 14:44:37 alarm bluetoothd[260]: Bluetooth daemon 5.48
Feb 19 14:44:37 alarm bluetoothd[260]: Starting SDP server
Feb 19 14:44:37 alarm bluetoothd[260]: kernel lacks bnep-protocol support
Feb 19 14:44:37 alarm bluetoothd[260]: System does not support network plugin
Feb 19 14:44:37 alarm bluetoothd[260]: Bluetooth management interface 1.14 initialized
Feb 19 14:44:37 alarm dbus-daemon[263]: [system] Activating via systemd: service name='org.freedesktop.hostname1' unit='dbus-org.freedesktop.hostname1.service' requested by ':1.4' (uid=0 pid=260 comm="/usr/lib/bluetooth/bluetoothd ")
Feb 19 14:44:37 alarm systemd[1]: bluetooth.service: Got notification message from PID 260 (STATUS=Starting up)
Feb 19 14:44:37 alarm systemd[1]: bluetooth.service: Got notification message from PID 260 (STATUS=Running)
Feb 19 14:44:37 alarm systemd[1]: bluetooth.service: Got notification message from PID 260 (READY=1)
Feb 19 14:44:37 alarm systemd[1]: bluetooth.service: D-Bus name org.bluez now registered by :1.4
Feb 19 14:44:37 alarm systemd[1]: bluetooth.service: Changed start -> running
Feb 19 14:44:37 alarm systemd[1]: bluetooth.service: Job bluetooth.service/start finished, result=done
Feb 19 14:44:37 alarm systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/bluetooth_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=13 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Feb 19 14:44:37 alarm systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/bluetooth_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=14 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Feb 19 14:44:37 alarm systemd-logind[256]: Got message type=signal sender=:1.3 destination=n/a path=/org/freedesktop/systemd1/unit/bluetooth_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=13 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Feb 19 14:44:37 alarm systemd-logind[256]: Got message type=signal sender=:1.3 destination=n/a path=/org/freedesktop/systemd1/unit/bluetooth_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=14 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Feb 19 14:44:37 alarm systemd[1]: bluetooth.target changed dead -> active
Feb 19 14:44:37 alarm systemd[1]: bluetooth.target: Job bluetooth.target/start finished, result=done
Feb 19 14:44:37 alarm systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/bluetooth_2etarget interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=23 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Feb 19 14:44:37 alarm systemd-logind[256]: Got message type=signal sender=:1.3 destination=n/a path=/org/freedesktop/systemd1/unit/bluetooth_2etarget interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=23 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Feb 19 14:44:38 alarm NetworkManager[281]: <info>  [1519051478.1463] Loaded device plugin: NMBluezManager (/usr/lib/NetworkManager/libnm-device-plugin-bluetooth.so)
Feb 19 14:44:40 alarm systemd[343]: sys-subsystem-bluetooth-devices-hci0.device: Changed dead -> plugged
Feb 19 14:44:40 alarm systemd[343]: sys-devices-platform-700b0200.sdhci-mmc_host-mmc1-mmc1:0001-mmc1:0001:3-bluetooth-hci0.device: Changed dead -> plugged
Feb 19 14:45:07 alarm systemd[404]: sys-subsystem-bluetooth-devices-hci0.device: Changed dead -> plugged
Feb 19 14:45:07 alarm systemd[404]: sys-devices-platform-700b0200.sdhci-mmc_host-mmc1-mmc1:0001-mmc1:0001:3-bluetooth-hci0.device: Changed dead -> plugged
Feb 19 14:45:15 alarm systemd[404]: obex.service: About to execute: /usr/lib/bluetooth/obexd
Feb 19 14:45:15 alarm systemd[404]: obex.service: Forked /usr/lib/bluetooth/obexd as 500
Feb 19 14:45:15 alarm systemd[500]: obex.service: Executing: /usr/lib/bluetooth/obexd
```

**Comment by vartom** [comment]
To work the bluetooth you need to connect the firmware( bcm4350c0.hcd) to it.
like that
sudo brcm_patchram_plus -d --patchram /system/etc/firmware/bcm4350с0.hcd --baudrate 3000000 --enable_lpm --enable_hci --use_baudrate_for_download --no2bytes --tosleep 1000 /dev/ttyTHS2

**Comment by denysvitali** [comment]
Thanks. I'll try to implement it :+1:

**Comment by denysvitali** [comment]
The Wi-Fi chip does work. The patches were upstreamed and are now available in the latest kernel release images (linux-smaug).
![img_20180220_154307](https://user-images.githubusercontent.com/4939519/36436469-9b772acc-1664-11e8-8b7c-1aaaf342a3a0.jpg)  
  
I'll open another issue for the Bluetooth problem

**Comment by denysvitali** [comment]
The bluetooth issue is now open at #10
